org.scala-ide.sdt.core/src/org/scalaide/core/internal/jdt/model/ScalaClassFileProvider.scala

```diff
@@ -11,19 +11,36 @@ import org.eclipse.jdt.internal.core.ClassFile
 import org.eclipse.jdt.internal.core.PackageFragment
 import org.scalaide.core.internal.project.ScalaProject
 
+/**
+ * Provides a `ScalaClassFile` implementation for classfiles that belong to
+ * Scala sources.
+ *
+ * This class caches the result of the association between package fragments
+ * (usually a jar) and classfiles. The cache is based on a heuristic because
+ * scanning the entire classpath may be expensive.
+ */
 class ScalaClassFileProvider extends IClassFileProvider with HasLogger {
 
-  /** @return a ScalaClassFile implementation if bytes represent a Scala classfile, or `null`
-   *          if the default JDT implementation should be used.
+  /**
+   * Returns a `ScalaClassFile` implementation if `contents` represent a Scala
+   * classfile or `null` if the default JDT implementation should be used.
    */
   override def create(contents: Array[Byte], parent: PackageFragment, name: String): ClassFile = {
     def updateCache(isScalaClassfile: Boolean): Unit = {
       val pfr = parent.getPackageFragmentRoot()
       if (pfr ne null)
         scalaPackageFragments.synchronized {
           if (!scalaPackageFragments.isDefinedAt(pfr)) {
-            logger.debug(s"Setting ${pfr.getElementName} (because of class $name) to be ${if (isScalaClassfile) "Scala" else "Java"}")
-            scalaPackageFragments += pfr -> isScalaClassfile
+            val jarName = pfr.getElementName
+            val isProbablyScalaArtifact = jarName matches """.*_2\.\d+.*"""
+            val ignoreClassfile = !isScalaClassfile && isProbablyScalaArtifact
```

Comment:
I thought about that. We could just look at the library name to find out if it is a Scala library. This way we wouldn't even need the cache anymore because we could immediately check in `isInteresting` if a library can contain Scala sources.

I didn't do this change so far because the current mechanism worked quite well. To me it looks like the algorithm described above (just looking at the library name) would work well but maybe I'm missing something. Can you think of a use case where it would not work?